sys/platform/src/lib.rs

```diff
@@ -0,0 +1,204 @@
+#![no_std]
+
+// Use std for tests
+#[cfg(test)]
+#[macro_use]
+extern crate std;
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+mod x86;
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+use x86::{self as cpu_id, Feature};
+
+#[cfg(all(target_arch = "aarch64", target_os = "macos"))]
+mod macos_arm;
+#[cfg(all(target_arch = "aarch64", target_os = "linux"))]
+mod linux_arm;
+
+#[cfg(test)]
+mod test;
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn simd128_support() -> bool {
+    cpu_id::supported(Feature::sse2)
+        && cpu_id::supported(Feature::sse3)
+        && cpu_id::supported(Feature::sse4_1)
+        && cpu_id::supported(Feature::avx)
+}
+
+#[cfg(all(target_arch = "aarch64", target_os = "linux"))]
+pub fn simd128_support() -> bool {
+    use linux_arm::*;
+    adv_simd()
+}
+
+#[cfg(all(target_arch = "aarch64", target_os = "macos"))]
+pub fn simd128_support() -> bool {
+    use macos_arm::*;
+    adv_simd()
+}
+
+#[cfg(all(target_arch = "aarch64", not(any(target_os = "linux", target_os = "macos"))))]
+pub fn simd128_support() -> bool {
+    false
+}
+
+#[cfg(not(any(target_arch = "x86", target_arch = "x86_64", target_arch = "aarch64")))]
+pub fn simd128_support() -> bool {
+    // XXX: Check for z14 or z15
+    false
+}
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn simd256_support() -> bool {
+    cpu_id::supported(Feature::avx2)
+}
+
+#[cfg(not(any(target_arch = "x86", target_arch = "x86_64")))]
+pub fn simd256_support() -> bool {
+    // XXX: Check for z14 or z15
+    false
+}
+
+pub fn x25519_support() -> bool {
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    return cpu_id::supported(Feature::bmi2) && cpu_id::supported(Feature::adx);
+
+    #[cfg(not(any(target_arch = "x86", target_arch = "x86_64")))]
+    false
+}
+
+#[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+pub fn bmi2_adx_support() -> bool {
+    cpu_id::supported(Feature::bmi2) && cpu_id::supported(Feature::adx)
+}
+
+#[cfg(not(any(target_arch = "x86", target_arch = "x86_64")))]
+pub fn bmi2_adx_support() -> bool {
+    false
+}
+
+/// Check whether p(cl)mull is supported
+pub fn pmull_support() -> bool {
```

Comment:
There are three different approaches in here for the guards
1. multiple functions, each guarded with the platform
2. one function with guards within the function
3. a module with the guards and all the functions

Each of the approaches has advantages and disadvantages. Please pick one and make all of them use the same approach.

Reply:
Accomplished with approach 2 for all functions.